go/pkg/casng/streaming_upload.go

```diff
@@ -188,10 +195,278 @@ func (u *uploader) streamPipe(ctx context.Context, in <-chan UploadRequest) <-ch
 
 // uploadBatcher handles files that can fit into a batching request.
 func (u *uploader) batcher() {
+	log.V(1).Info("[casng] upload.batcher.start")
+	defer log.V(1).Info("[casng] uploader.batcher.stop")
+
+	bundle := make(uploadRequestBundle)
+	bundleSize := u.uploadRequestBaseSize
+	ctx := u.ctx // context with unified metadata.
+
+	handle := func() {
+		if len(bundle) < 1 {
+			return
+		}
+		// Block the batcher if the concurrency limit is reached.
+		startTime := time.Now()
+		if !u.uploadThrottler.acquire(u.ctx) {
+			// Ensure responses are dispatched before aborting.
+			for d, item := range bundle {
+				u.dispatcherResCh <- UploadResponse{
+					Digest: d,
+					Stats:  Stats{BytesRequested: d.Size},
+					Err:    context.Canceled,
+					tags:   item.tags,
+					reqs:   item.reqs,
+				}
+			}
+			return
+		}
+		log.V(3).Infof("[casng] upload.batcher.throttle.duration; start=%d, end=%d", startTime.UnixNano(), time.Now().UnixNano())
+
+		u.workerWg.Add(1)
+		go func(ctx context.Context, b uploadRequestBundle) {
+			defer u.workerWg.Done()
+			defer u.uploadThrottler.release()
+			// TODO: cancel ctx if all requesters have cancelled their contexts.
+			u.callBatchUpload(ctx, b)
+		}(ctx, bundle)
+
+		bundle = make(uploadRequestBundle)
+		bundleSize = u.uploadRequestBaseSize
+		ctx = u.ctx
```

Comment:
why assigning to `u.ctx` again?

Reply:
To reset `ctx` now that the associated bundle has been dispatched with a copy of it.